cli: Add route config

lmars opened the pull request:
This updates the CLI to support a route config file that is capable of setting routes across apps in a single, atomic operation.

The route config file uses the [Starlark language](https://github.com/bazelbuild/starlark) to define a set of routes that should exist for a list of apps:

```starlark
def routes(ctx):
  return {
    "app1": [
      http_route(
        domain  = "www.example.com",
        service = "app1-web",
      ),
      http_route(
        domain  = "admin.example.com",
        service = "app1-admin-web",
      ),
    ],
    "app2": [
      http_route(
        domain  = "www.example.net",
        service = "app2-web",
      ),
    ],
    "app3": [
      tcp_route(
        port    = 2222,
        service = "app3-ssh",
      ),
    ],
  }
```

The route config is applied by running `flynn route config apply <file>`, which prints the changes that will be made and prompts the user for confirmation before making them:

```
$ flynn route config apply routes.cfg 

!!! You are about to make the following 3 route changes: !!!

+--------+-----------------------------+----------+--------------------------------+
| ACTION |            ROUTE            | SERVICE  |             CONFIG             |
+--------+-----------------------------+----------+--------------------------------+
| DELETE | http:app2.1.localflynn.com/ | app2-web | leader=false sticky=false      |
|        |                             |          | drain_backends=true            |
|        |                             |          | disable_keep_alives=false      |
| UPDATE | http:app3.1.localflynn.com/ | app3-web | leader=false sticky=false      |
|        |                             |          | drain_backends=true            |
|        |                             |          | disable_keep_alives=false      |
| CREATE | http:app2.1.localflynn.com/ | app1-web | leader=false sticky=false      |
|        |                             |          | drain_backends=true            |
|        |                             |          | disable_keep_alives=false      |
+--------+-----------------------------+----------+--------------------------------+

Are you sure you want to proceed? (yes/no): yes
The confirmed route changes were made successfully.
```

Route config for existing apps can be generated with `flynn route config generate <apps>...`:

```
$ flynn route config generate app1 app2 app3 > routes.cfg

$ cat routes.cfg
def routes(ctx):
  return {
    "app1": [
      http_route(
        domain  = "app1.1.localflynn.com",
        service = "app1-web",
      ),
    ],
    "app2": [
      http_route(
        domain  = "app2.1.localflynn.com",
        service = "app2-web",
      ),
    ],
    "app3": [
      http_route(
        domain  = "app3.1.localflynn.com",
        service = "app3-web",
      ),
    ],
  }

$ flynn route config apply routes.cfg 
No route changes were made as all routes are up-to-date.
```

I'm opening this PR early for feedback, although there are still a few tasks remaining:

- [x] improve the CLI dry run prompt to show a diff for route updates
- [x] add tests for `routeconfig.Generate` and `routeconfig.Load`
- [x] add tests for `flynn route config generate` and `flynn route config apply`
- [x] document the accepted route config syntax and options (probably at the top of the generated file)
- [x] support all route flags (i.e. leader, stickiness, drain_backends, disable_keep_alives)
- [x] support static certificates

jvatic commented on controller/routes_test.go:
```diff
@@ -347,3 +349,234 @@ func (s *S) TestStreamRouteEvents(c *C) {
+			c.Fatalf("expected %d routes, got %d", len(expected), len(routes))
+		}
+		for i, r := range expected {
+			c.Assert(routes[i].Service, Equals, r.ServiceTarget.ServiceName)
```
Should `DrainBackends` also be compared here?

lmars replied:
Handled in cebac8ed.

jvatic commented on controller/grpc.go:
```diff
@@ -951,6 +952,52 @@ func (g *grpcAPI) SetRoutes(ctx context.Context, req *api.SetRoutesRequest) (*ap
 
+func (g *grpcAPI) ListAppRoutes(ctx context.Context, req *api.ListAppRoutesRequest) (*api.ListAppRoutesResponse, error) {
+	if len(req.Apps) == 0 {
+		return nil, hh.ValidationErr("apps", "must be set")
```
Should this and the other errors returned be `grpc.Error`s?

lmars replied:
Errors refactored in aa4ceb32.

jvatic commented on controller/grpc.go:
```diff
@@ -938,6 +939,18 @@ func (g *grpcAPI) CreateDeployment(req *api.CreateDeploymentRequest, ds api.Cont
+func (g *grpcAPI) SetRoutes(ctx context.Context, req *api.SetRoutesRequest) (*api.SetRoutesResponse, error) {
+	changes, state, err := g.routeRepo.Set(req.AppRoutes, req.DryRun, req.ExpectedState)
+	if err != nil {
+		return nil, err
```
Should this error be converted to a `grpc.Error`?

lmars replied:
Errors refactored in aa4ceb32.

jvatic commented on cli/route.go:
```diff
@@ -38,15 +45,23 @@ Options:
 
+        $ flynn route config generate -f routes.cfg app1 app2 app3
+
+        $ flynn route config apply routes.cfg
```
The indentation appears to be a tiny but off for these compared to the other examples below

lmars replied:
Good spot 👀 Fixed in 8b4c8a59.

jvatic commented on cli/route.go:
```diff
@@ -118,6 +142,161 @@ func runRoute(args *docopt.Args, client controller.Client) error {
+	apps := args.All["<apps>"].([]string)
+	req := api.ListAppRoutesRequest{Apps: apps}
+	var res api.ListAppRoutesResponse
+	if err := client.Invoke("flynn.api.v1.Router/ListAppRoutes", &req, &res); err != nil {
```
I assume this will be changed to use the gRPC protocol via the generated client vs gRPC-web once that is exposed?

lmars replied:
Yes, although adding gRPC support to the router will come in another PR, for now we'll use gRPC web.

titanous commented on pkg/routeconfig/routeconfig.go:
```diff
@@ -0,0 +1,144 @@
+
+	// load the config using skycfg
+	ctx := context.Background()
+	config, err := skycfg.Load(ctx, tmp.Name())
```
Use a `skycfg.FileReader` that reads from memory instead of writing to a tempfile.

lmars replied:
Done in 151ec5a1 (and refactored in 77cab2b).

lmars commented:
I've made a bunch of changes to address the initial TODOs and comments added.

The changes are largely untested but I wanted to get feedback on the approach before adding more tests.

### Config Structure

I've updated the structure of the config file to be more self contained. Rather than appending strings, the route config is a full skycfg file with `main()` expected to return a `RouteConfig` protocol message with an appropriate version (allowing us to iterate on versions of the route config in the future).

I've also added comments to the file, so a generated route config now looks like this:

https://gist.github.com/lmars/10856e64110f1f89a6a66a99343affca

### Improved diffs

I've improved the output of `flynn route config apply` to better communicate changes by showing both the before and after routes:

```
$ flynn route config apply routes.cfg 

!!! You are about to make the following 3 route changes: !!!

+--------+-----------------------------+--------------------------------------+--------------------------------------+
| ACTION |            ROUTE            |                BEFORE                |                AFTER                 |
+--------+-----------------------------+--------------------------------------+--------------------------------------+
| DELETE | http:app2.1.localflynn.com/ | service: app-web                     | -                                    |
+--------+-----------------------------+--------------------------------------+--------------------------------------+
| UPDATE | http:app.1.localflynn.com/  | service: app-web                     | service: app-web (leader)            |
|        |                             | certificate: app_1_localflynn_com_v1 | certificate: app_1_localflynn_com_v1 |
+--------+-----------------------------+--------------------------------------+--------------------------------------+
| CREATE | http:app3.1.localflynn.com/ | -                                    | service: app-web                     |
+--------+-----------------------------+--------------------------------------+--------------------------------------+
```

### Static Certificates

I've added support for managing static certificates using the CLI and also referencing them in the route config file.

The new commands are:

```
$ flynn certificate

$ flynn certificate add [--ref=REF] [--meta=KEY=VAL...] CERT KEY

$ flynn certificate inspect NAME

$ flynn certificate remove NAME
```

Each certificate has a "ref" that can be used to reference it in the route config file, with the system certificate having the ref `flynn_default_certificate`, and user supplied certificates being assigned refs like `<common-name>_v<n>` (e.g. `app_example_com_v1`) which can be overridden using the `--ref` flag to `flynn certificate add`.

For example, after bootstrapping a local dev cluster:

- list certificates:

```
$ flynn certificate
+-------------------------------------------+--------------------------------------+--------------------+-------------------------+
|                    REF                    |                ROUTES                |      DOMAINS       |          META           |
+-------------------------------------------+--------------------------------------+--------------------+-------------------------+
| certificates/flynn_default_certificate    | 275d5172-c8fd-4ba9-9c47-2bf4e0593ffa | 1.localflynn.com   | flynn.ephemeral_ca=true |
|                                           | 5a8bfc6d-a0ba-4c79-b12f-83b86ac2f849 | *.1.localflynn.com |                         |
|                                           | 8306b851-9967-429d-8e5d-aee302febd62 |                    |                         |
|                                           | 733b14d9-e0d0-4f3b-a8ae-234836a56860 |                    |                         |
+-------------------------------------------+--------------------------------------+--------------------+-------------------------+
```

- add a new certificate:

```
$ openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem -subj '/CN=app.1.localflynn.com' -addext "subjectAltName = DNS:app.1.localflynn.com" -days 365 -nodes

$ flynn certificate add cert.pem key.pem
certificates/app_1_localflynn_com_v1
```

- inspect the new certificate:

```
$ flynn certificate inspect certificates/app_1_localflynn_com_v1
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 220798855510188520569757294941507424465197705231 (0x26acf68e4cf6452bf1561c532b2c000f89c7a80f)
    Signature Algorithm: SHA256-RSA
        Issuer: CN=app.1.localflynn.com
        Validity
            Not Before: Feb 11 14:11:24 2020 UTC
            Not After : Feb 10 14:11:24 2021 UTC
        Subject: CN=app.1.localflynn.com
...
        X509v3 extensions:
            X509v3 Subject Key Identifier:
                0C:07:64:FA:62:FE:36:BC:CD:02:4F:5D:5E:75:49:EE:B1:85:6C:F1
            X509v3 Authority Key Identifier:
                keyid:0C:07:64:FA:62:FE:36:BC:CD:02:4F:5D:5E:75:49:EE:B1:85:6C:F1
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Alternative Name:
                DNS:app.1.localflynn.com

```

- update route config to reference the new certificate and apply:

```
def routes(ctx):
  return {
    "app": [
      http_route(
        domain = "app.1.localflynn.com",
        target = service("app-web"),
        certificate = "certificates/app_1_localflynn_com_v1",
      ),
    ],
  }
```

```
$ flynn route config apply routes.cfg 

!!! You are about to make the following route change: !!!

+--------+----------------------------+------------------+--------------------------------------+
| ACTION |           ROUTE            |      BEFORE      |                AFTER                 |
+--------+----------------------------+------------------+--------------------------------------+
| UPDATE | http:app.1.localflynn.com/ | service: app-web | service: app-web                     |
|        |                            |                  | certificate: app_1_localflynn_com_v1 |
+--------+----------------------------+------------------+--------------------------------------+

Are you sure you want to proceed? (yes/no): yes
The confirmed route changes were made successfully.
```

- see that the router now serves the certificate for `app.1.localflynn.com`:

```
$ curl -v --cacert cert.pem https://app.1.localflynn.com 2>&1 | grep -A6 -F 'Server certificate'
* Server certificate:
*  subject: CN=app.1.localflynn.com
*  start date: Feb 11 14:11:24 2020 GMT
*  expire date: Feb 10 14:11:24 2021 GMT
*  subjectAltName: host "app.1.localflynn.com" matched cert's "app.1.localflynn.com"
*  issuer: CN=app.1.localflynn.com
*  SSL certificate verify ok.
```

@titanous @jvatic PTAL (I will add some more tests once you've reviewed, then will ask for a final review).

lmars commented on controller/data/route.go:
```diff
@@ -620,3 +847,94 @@ func normaliseRoutePath(path string) string {
+func routesEqualForUpdate(existing *router.Route, desired *api.Route) bool {
+	// check whether they have the same service configuration
+	if desired.ServiceTarget == nil {
+		return existing.Service == ""
```
Handled in cebac8ed.

lmars commented on cli/route.go:
```diff
@@ -5,20 +5,27 @@ import (
 func init() {
 	register("route", runRoute, `
 usage: flynn route
+       flynn route config generate [-f <file>] <apps>...
```
Done in d5c01439.

lmars commented on controller/api/router.proto:
```diff
@@ -0,0 +1,156 @@
+message ListAppRoutesResponse {
+  // app_routes are the requested routes.
+  repeated AppRoutes app_routes = 1;
+}
```
I did add this, but then after adding certificate management, the state now depends on certificates too, so I dropped the change. Do you think it should still be added?

lmars replied:
Certificates are now identified by fingerprints, so I've re-instated the original change to add state here.

lmars commented on controller/data/route.go:
```diff
@@ -34,41 +31,266 @@ func NewRouteRepo(db *postgres.DB) *RouteRepo {
+		return nil, nil, err
+	}
+
+	existingRoutes, err := r.list(tx, "")
```
Done in 9db8e17a.

lmars commented on controller/data/route.go:
```diff
@@ -34,41 +31,266 @@ func NewRouteRepo(db *postgres.DB) *RouteRepo {
+}
+
 func (r *RouteRepo) Add(route *router.Route) error {
+	existingRoutes, err := r.List("")
```
Done in 9db8e17a.

lmars commented on controller/data/route.go:
```diff
@@ -313,10 +513,26 @@ func (r *RouteRepo) listTCP(parentRef string) ([]*router.Route, error) {
 }
 
 func (r *RouteRepo) Update(route *router.Route) error {
+	existingRoutes, err := r.List("")
```
Done in 9db8e17a.

lmars commented on controller/data/route.go:
```diff
@@ -392,10 +602,26 @@ func (r *RouteRepo) updateTCP(tx *postgres.DBTx, route *router.Route) error {
 }
 
 func (r *RouteRepo) Delete(route *router.Route) error {
+	existingRoutes, err := r.List("")
```
Done in 9db8e17a.

lmars commented on controller/data/route.go:
```diff
@@ -34,41 +31,266 @@ func NewRouteRepo(db *postgres.DB) *RouteRepo {
+		app, err := selectApp(r.db, appID, false)
+		if err != nil {
+			if err == ErrNotFound {
+				err = fmt.Errorf("app not found: %v", appID)
```
Changed to a validation error in aa4ceb32.

lmars commented on controller/data/route.go:
```diff
@@ -437,3 +655,286 @@ func (r *RouteRepo) createEvent(tx *postgres.DBTx, route *router.Route, typ ct.E
+		if existing.Type != "http" || (op == routeOpUpdate && existing.ID == route.ID) {
+			continue
+		}
+		if existing.Domain == route.Domain && existing.Port == route.Port && existing.Path == route.Path {
```
I've kept the unique index: 0ae9e566.

lmars commented on pkg/routeconfig/routeconfig.go:
```diff
@@ -0,0 +1,144 @@
+const configMain = `
+api = proto.package("flynn.api.v1")
+
+def main(ctx):
```
This is now versioned by having the route config be a self contained skycfg file, with the only expectation that it returns a single protocol message. For now, we expect it t be a `flynn.api.v1.RouteConfig` with a version of `RouteConfig_VERSION_1`, but in future we can update the CLI to handle the return of other top level objects as a way to migrate.

See 77cab2b2.

WDYT? I guess it's hard to predict _how_ we might change this file in the future.

titanous replied:
How about we put all of the helpers in a versioned import? That way we don't have tons of boilerplate, can iterate on the helpers, and complicated stuff can opt out.

lmars replied:
Done, config now looks like this:

```starlark
load("flynn.routeconfig.v1", "config")

def main(ctx):
  return config.app_routes({
    "app": [
      config.http_route(
        domain = "app.1.localflynn.com",
        target = config.service("app-web"),
      ),
    ],
  })
```

titanous commented on controller/api/api.go:
```diff
@@ -712,3 +753,168 @@ func NewJobState(from ct.JobState) DeploymentEvent_JobState {
+	}
+
+	// set Issuer and validity times
+	cert.Issuer = leafCert.Issuer.String()
```
The chain should be checked to make sure:

- that the issuer always matches the subject of the next certificate in the chain exactly
- the CA attribute is set on all except the leaf
- the `serverAuth` or `anyExtendedKeyUsage` EKU is set on all certificates in the chain

lmars replied:
I added the first two:

https://github.com/flynn/flynn/blob/cda81f4dda6cfcc758ee2586d81310fd44579238/controller/api/api.go#L845-L867

but can you clarify the third: doesn't `serverAuth` just need to be set on the leaf cert?

titanous replied:
In order for the chain to be valid for a given EKU, that EKU must validate through on each certificate in the chain. This means each certificate in the chain must have the specific EKU, `anyExtendedKeyUsage` (effectively a wildcard), or no EKU extension at all (for backwards compatibility, most roots don't have an EKU). So the validation is:

- The leaf must have `serverAuth`
- Each certificate in the chain must either have no EKU, `serverAuth`, or `anyExtendedKeyUsage`

titanous replied:
Also, add one last validation: all signatures in the chain must use a SHA-2 hash algorithm.

lmars replied:
Ok thanks for clarifying, updated in b25ea39e.

titanous commented on controller/grpc.go:
```diff
@@ -998,9 +1006,73 @@ outer:
+		}
+	}
+	var keyPEM bytes.Buffer
+	if err := pem.Encode(&keyPEM, &pem.Block{Type: "PRIVATE KEY", Bytes: staticCert.PrivateKey}); err != nil {
```
The certificate chain shouldn't be PEM-encoded in the database, it should be stored as an array of DER binary certificates. The key should also just be DER binary.

lmars replied:
Done in 3bf6384b.

lmars commented:
After playing around with certificate management using the route config file and considering review comments, I've refactored the last change I made:

- the helper methods are now loaded into a `config` variable rather than appearing inline, so the config is simplified to:

```starlark
load("flynn.routeconfig.v1", "config")

def main(ctx):
  return config.app_routes({
    "app": [
      config.http_route(
        domain = "app.1.localflynn.com",
        target = config.service("app-web"),
      ),
    ],
  })
```

- I've moved private keys out into their own db table so they can be added independently of certificates, and added a `flynn tls-key` command to manage them:

```bash
$ flynn tls-key
$ flynn tls-key add <key>
$ flynn tls-key remove <name> 
```

- I've dropped the `flynn certificate` command and the ref changes in favour of specifying static certificates directly in the config file:

```starlark
load("flynn.routeconfig.v1", "config")

def main(ctx):
  return config.app_routes({
    "app": [
      config.http_route(
        domain = "app.1.localflynn.com",
        target = config.service("app-web"),
        certificate = cert,
      ),
    ],
  })

cert = config.static_certificate('''
-----BEGIN CERTIFICATE-----
MIIFQDCCAyigAwIBAgIUYuDdusMZWAHIu5Yo9B3p9UnDrv8wDQYJKoZIhvcNAQEL
BQAwHzEdMBsGA1UEAwwUYXBwLjEubG9jYWxmbHlubi5jb20wHhcNMjAwMjIxMDMy
MzIzWhcNMjEwMjIwMDMyMzIzWjAfMR0wGwYDVQQDDBRhcHAuMS5sb2NhbGZseW5u
...
uUjr1QImCEyKqZOeKfHw4gC3wnhPQoihvrYBaBcnY5opVPsnqDV1EJ590CCBIe6g
1OJ5R9ReGVG6Nl8XggBPJ76AmgvtFeqzmLCnAtJRNN0XoLe4gHpBVIew7QMLGFiT
WnUX2A==
-----END CERTIFICATE-----
''')
```

- both TLS keys and certificates are identified with fingerprints, meaning the expected key ID for a given static certificate can be generated rather than explicitly defined in the route config, and dry runs can report certificate ID changes without actually changing certificates in the database:

```bash
# applying with a new certificate fails as the key is missing
$ flynn route config apply routes.cfg 
error applying config: unexpected Grpc-Status: 3 (certificate key not found: d71ecc1f71fa8c6f79b6824107a62b272e8bcd15ec807ecfcbeb71405cd219bb)

# add the key
$ flynn tls-key add key.pem 
tls-keys/d71ecc1f71fa8c6f79b6824107a62b272e8bcd15ec807ecfcbeb71405cd219bb

# applying now succeeds
$ flynn route config apply routes.cfg 

!!! You are about to make the following route change: !!!

+--------+----------------------------+------------------+-------------------------------------------------------------------------------+
| ACTION |           ROUTE            |      BEFORE      |                                     AFTER                                     |
+--------+----------------------------+------------------+-------------------------------------------------------------------------------+
| UPDATE | http:app.1.localflynn.com/ | service: app-web | service: app-web                                                              |
|        |                            |                  | certificate: e041742c84afe31ec3cc05a73038183569bee45af9a63ea658b05e653229d923 |
+--------+----------------------------+------------------+-------------------------------------------------------------------------------+

Are you sure you want to proceed? (yes/no): yes
The confirmed route changes were made successfully.
```

I'll also respond to individual comments, and as with my last update, I'll add tests once we agree on the approach.

lmars commented on cli/certificate.go:
```diff
@@ -0,0 +1,158 @@
+usage: flynn certificate
+       flynn certificate add [--ref=<ref>] [--meta=<key>=<val>...] <cert> <key>
+       flynn certificate remove <name>
+       flynn certificate inspect <name>
```
This command has been dropped in favour of specifying static certificates in the route config file (see https://github.com/flynn/flynn/pull/4581#issuecomment-590100925)

lmars commented on cli/certificate.go:
```diff
@@ -0,0 +1,158 @@
+func init() {
+	register("certificate", runCertificate, `
+usage: flynn certificate
+       flynn certificate add [--ref=<ref>] [--meta=<key>=<val>...] <cert> <key>
```
I decided to just drop it, since as you say it will seldom be used anyway.

lmars commented on controller/api/api.go:
```diff
@@ -712,3 +753,168 @@ func NewJobState(from ct.JobState) DeploymentEvent_JobState {
+	// set the Domains from SAN values, falling back to the Common Name
+	// from the Subject
+	cert.Domains = leafCert.DNSNames
+	if len(cert.Domains) == 0 {
```
Done: https://github.com/flynn/flynn/blob/cda81f4dda6cfcc758ee2586d81310fd44579238/controller/api/api.go#L837-L843

lmars commented on controller/api/api.go:
```diff
@@ -712,3 +753,168 @@ func NewJobState(from ct.JobState) DeploymentEvent_JobState {
+	tlsCert, err := tls.X509KeyPair([]byte(from.Cert), []byte(from.Key))
+	if err != nil {
+		cert.Status = StaticCertificate_STATUS_INVALID
+		cert.StatusDetail = err.Error()
```
Done: https://github.com/flynn/flynn/blob/cda81f4dda6cfcc758ee2586d81310fd44579238/controller/data/route.go#L943-L944

lmars commented on controller/api/router.proto:
```diff
@@ -63,6 +81,106 @@ message ListAppRoutesResponse {
+    // chain from client platform trusted roots to the leaf. If force is set to
+    // false, a variety of sanity checks are performed, the chain can be out of
+    // order, and it will be checked and rebuilt for client platform
+    // compatibility.
```
Done: https://github.com/flynn/flynn/blob/cda81f4dda6cfcc758ee2586d81310fd44579238/controller/api/router.proto#L298-L303

lmars commented on controller/data/route.go:
```diff
@@ -319,34 +339,105 @@ func (r *RouteRepo) addTCP(tx *postgres.DBTx, route *router.Route) error {
+
+	// fallback to the current date
+	if cn == "" {
+		cn = time.Now().Format("20060102")
```
This has been dropped in favour of fingerprints for identifying certificates.

lmars commented on controller/data/route.go:
```diff
@@ -319,34 +339,105 @@ func (r *RouteRepo) addTCP(tx *postgres.DBTx, route *router.Route) error {
+func genCertRef(db dbOrTx, cert, key string) string {
+	// determine the Subject Common Name
+	cn := func() string {
+		c, err := tls.X509KeyPair([]byte(cert), []byte(key))
```
This has been dropped, and certificates and keys are now stored as DER-encoded bytes in the database: 3bf6384b

lmars commented on controller/data/route.go:
```diff
@@ -636,6 +756,121 @@ func (r *RouteRepo) deleteTx(tx *postgres.DBTx, route *router.Route) error {
+	// ensure the certificate is not referenced by any routes
+	if len(cert.Routes) > 0 {
+		tx.Rollback()
+		return nil, fmt.Errorf("cannot delete certificate as it is referenced by the following routes: %s", strings.Join(cert.Routes, ", "))
```
This code has been dropped since I dropped the `flynn certificate` command

lmars commented on controller/data/route.go:
```diff
@@ -319,34 +339,105 @@ func (r *RouteRepo) addTCP(tx *postgres.DBTx, route *router.Route) error {
 			return err
 		}
-		if err := tx.Exec("route_certificate_insert", rid, cert.ID); err != nil {
+		if err := db.Exec("route_certificate_insert", rid, cert.ID); err != nil {
```
This change has been reverted.

titanous commented on router/types/types.go:
```diff
@@ -1,24 +1,366 @@
+func (c *Certificate) MarshalJSON() ([]byte, error) {
+	key := c.redactJSONKey
+	if key == "" {
+		key = c.KeyPEM()
```
The key field should never be exposed via the API (except the redacted version that shows up in the logs to indicate that the field was set).

lmars replied:
The reason the key was in the JSON was because that's how the router would get it (specifically from a controller route event that contains the JSON encoded route + certificate).

I've replaced the key with the keyID in the JSON and added a separate `GetPrivateKey` gRPC method that the router uses to get the key (which I guess we can ultimately lock down further so only the router can call that method).

See e1905c2b.

lmars commented on controller/data/route.go:
```diff
@@ -355,6 +371,101 @@ func (r *RouteRepo) addRouteCertWithTx(tx *postgres.DBTx, route *router.Route) e
+	}
+
+	// get the key
+	key, err := scanKey(tx.QueryRow("tls_key_select", strings.TrimPrefix(name, "tls-keys/")))
```
Updated in 591c973b.

lmars commented on controller/data/route.go:
```diff
@@ -809,20 +922,37 @@ func (r *RouteRepo) validateHTTP(route *router.Route, existingRoutes []*router.R
+			}
+		} else {
+			keyID := cert.KeyID()
+			if _, err := scanKey(r.db.QueryRow("tls_key_select", keyID)); err != nil {
```
Updated in 591c973.

lmars commented on router/types/types.go:
```diff
@@ -1,24 +1,366 @@
+// ID returns the unique ID of this Certificate
+func (c *Certificate) ID() string {
+	digest := sha256.Sum256(bytes.Join(c.Chain, []byte{}))
+	return hex.EncodeToString(digest[:])
```
Yes I actually made that change then dropped it for some reason, re-instated in 24481e26.

lmars commented on router/types/types.go:
```diff
@@ -1,24 +1,366 @@
+	}
+	chain := make([]string, len(c.Chain))
+	for i, cert := range c.Chain {
+		chain[i] = string(pem.EncodeToMemory(&pem.Block{
```
Done in b5bb4538.

lmars commented on controller/data/schema.go:
```diff
@@ -1115,3 +1118,161 @@ func migrateProcessData(tx *postgres.DBTx) error {
+INSERT INTO tls_key_algorithms (name) VALUES ('ecc-p256'), ('rsa-2048'), ('rsa-4096');
+
+CREATE TABLE tls_keys (
+  id         text        PRIMARY KEY,
```
Done in f509e498.

lmars commented on router/types/types.go:
```diff
@@ -1,24 +1,366 @@
+	if key, err := x509.ParseECPrivateKey(der); err == nil {
+		return key, nil
+	}
+	return nil, errors.New("failed to parse private key (tried PKCS#1, PKCS#8 and SEC1)")
```
Done in 72a7d841.

lmars commented on cli/route.go:
```diff
@@ -118,6 +142,190 @@ func runRoute(args *docopt.Args, client controller.Client) error {
+	if len(apps) == 0 {
+		app, err := app()
+		if err != nil {
+			return errors.New("no app found, run from a repo with a flynn remote, specify one with -a or pass a list of apps on the command line")
```
Done in 72a7d841.

lmars commented on pkg/routeconfig/routeconfig.go:
```diff
@@ -102,26 +105,30 @@ var configTemplate = template.Must(template.New("routes.cfg").Parse(`
+# that represent the routes that should exist for a set of apps.
 
-apiv1 = proto.package("flynn.api.v1")
+# load the v1 config helpers
```
Removed in 56e40cd7.

lmars commented on pkg/routeconfig/routeconfig.go:
```diff
@@ -102,26 +105,30 @@ var configTemplate = template.Must(template.New("routes.cfg").Parse(`
-# each app.
-def routes(ctx):
-  return {
+# return the routes
```
Removed in 56e40cd7.

lmars commented on pkg/routeconfig/routeconfig.go:
```diff
@@ -0,0 +1,272 @@
+# To ensure the routes defined in this file exist (and that routes not defined in this file don't
+# exist), apply it by running:
+#
+#     flynn route config apply path/to/routes.cfg
```
Yes that makes sense, updated in 56e40cd7.

lmars commented on pkg/routeconfig/routeconfig.go:
```diff
@@ -0,0 +1,272 @@
+# and is processed using the Skycfg extension library (https://github.com/stripe/skycfg).
+#
+# A 'main' function must be defined that returns a list of flynn.api.v1.AppRoutes protocol messages
+# that represent the routes that should exist for a set of apps.
```
I've actually documented all the methods in the generated config in 28cc62f4.

lmars commented:
Assuming this is pretty much the design we want, I'm going to add some tests.

lmars commented:
Test added for `pkg/routeconfig` (d6cab8a0) and `flynn route config apply` (e77afe3a).

@titanous this is ready for another review.

titanous commented:
Flynn is unmaintained and our infrastructure will shut down on June 1, 2021. See [the README](https://github.com/flynn/flynn#flynn-is-unmaintained) for details.
